### PR TITLE
Test logic

### PR DESCRIPTION
**Pull request by bgy36ww**
Update Test logic to dynamic determine results based on list operations

### CODE REVIEW

File: dab/system.py [hunk 1]
```diff
@@ -13,24 +18,45 @@ def restart(test_result, durationInMs=0,expectedLatencyMs=0):
...
+    except Exception as error:
+        print("Schema error:", error)
+        return False
     response  = jsons.loads(test_result.response)
-    if response['status'] != 200:
+    request = jsons.loads(test_result.request)
+    if not EnforcementManager().is_setting_supported(request.system_setting_key):
+        if response['status'] != 501:
+            return False
+    elif response['status'] != 200:
```
Comment (xhaakon) [inline dab/system.py:31]: `EnforcementManager().is_setting_supported(request.system_setting_key)` fails with `AttributeError: 'dict' object has no attribute 'system_setting_key'`.

If I get [DAB 2.0 specification of `/system/settings/get`](https://github.com/device-automation-bus/dab-specification-2.0/blob/main/DAB.md#operation-getting-system-settings) right, you don't specify which setting(s) you want to get and you always retrieve _all_ settings values in the reply. You'll therefore never see a 501 response and it is pointless to check for it. I suggest reverting this change.
```suggestion
    if response['status'] != 200:
```

**Review by xhaakon — CHANGES_REQUESTED** [review]

File: util/enforcement_manager.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        return not self.supported_keys or key in self.supported_keys
+    
+    def add_supported_voice_assistant(self, voice_assistant):
+        self.supported_voice_assistants.add(voice_assistant)
+
+    def is_voice_assistant_supported(self, voice_assistant):
+        return not self.supported_voice_assistants or voice_assistant in self.supported_voice_assistants
+    
+    def get_voice_assistant(self):
+        return "Alexa" if len(self.supported_voice_assistants) == 0 else self.supported_voice_assistants[0]
```
Comment (xhaakon) [inline util/enforcement_manager.py:54]: The name should be `AmazonAlexa` according to [DAB 2.0](https://github.com/device-automation-bus/dab-specification-2.0/blob/main/DAB.md#voice-system-name-registry).
```suggestion
        return "AmazonAlexa" if len(self.supported_voice_assistants) == 0 else self.supported_voice_assistants[0]
```

**Review by xhaakon — CHANGES_REQUESTED** [review]

**Comment by bgy36ww** [comment]
Hey Jakub Adam,

Thanks for the review, I caught COVID. And couldn't get back to this PR.

File: util/enforcement_manager.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        """
+
+        if not self.supported_settings:
+            return True
+
+        if isinstance(getattr(self.supported_settings, setting), List) and len(getattr(self.supported_settings, setting)) == 0:
+            return True
+
+        if isinstance(getattr(self.supported_settings, setting), bool) and getattr(self.supported_settings, setting):
+            return True
```
Comment (xhaakon) [inline util/enforcement_manager.py:77]: I'm getting `AttributeError: 'dict' object has no attribute 'language'` here.

```suggestion
        if isinstance(self.supported_settings.get(setting), List) and len(self.supported_settings.get(setting)) == 0:
            return True

        if isinstance(self.supported_settings.get(setting), bool) and self.supported_settings.get(setting):
            return True
```

File: voice_text.py [hunk 4]
```diff
@@ -1,16 +1,18 @@
...
-    ("voice/send-text",'{"requestText" : "Pause Video", "voiceSystem": "Alexa"}', dab.voice.send_text, "Did video paused?", "voice pause"),
-    ("voice/send-text",'{"requestText" : "Fast forward video", "voiceSystem": "Alexa"}', dab.voice.send_text, "Did video playback fast forward?", "voice fastforward"),
-    ("voice/send-text",'{"requestText" : "Rewind video", "voiceSystem": "Alexa"}', dab.voice.send_text, "Did video playback rewind?", "voice rewind"),
-    ("voice/send-text",'{"requestText" : "Exit to main menu", "voiceSystem": "Alexa"}', dab.voice.send_text, "Are you on main menu?", "voice exit"),
+    ("voice/list",'{}', dab.voice.list, 200, "Voice List"),
+    ("voice/send-text",f'{{"requestText" : "Play lady Gaga music on YouTube", "voiceSystem": "{EnforcementManager().get_voice_assistant()}"}}', dab.voice.send_text, "Are you on search page with Lady Gaga?", "Voice launch Lady gaga"),
+    ("voice/send-text",f'{{"requestText" : "Press enter", "voiceSystem": "{EnforcementManager().get_voice_assistant()}"}}', dab.voice.send_text, "Is video playing?", "Voice play video", "Voice play video"),
```
Comment (xhaakon) [inline voice_text.py:10]: There seems to be redundant value at the end of the tuple causing this error when I try to list available test cases:

```
  File "/home/collabora/ws/GOO/dab-compliance-suite/main.py", line 85, in <module>
    (dab_request_topic, dab_request_body, validate_output_function, expected_response, test_title) = test_case
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: too many values to unpack (expected 5)
```

```suggestion
    ("voice/send-text",f'{{"requestText" : "Press enter", "voiceSystem": "{EnforcementManager().get_voice_assistant()}"}}', dab.voice.send_text, "Is video playing?", "Voice play video"),
```

**Review by xhaakon — CHANGES_REQUESTED** [review]
Hello Wei, I'm sorry to hear about your health issues. Hope you're already feeling better.

I found a couple more issues with this pull request.

**Comment by xhaakon** [comment]
On top of that validation of `/system/settings/set` responses is failing:

```
testing system/settings/set   {"memc": true} ... [ 
Error: self.dab_client.last_error_code()
Request invalid or malformed ]

testing system/settings/set   {"cec": true} ... Schema error: 'system_setting_key' is a required property

Failed validating 'required' in schema:
    {'properties': {'error': {'type': ['string', 'null']},
                    'status': {'type': 'integer'},
                    'system_setting_key': {'type': 'string'},
                    'value': {}},
     'required': ['status', 'system_setting_key', 'value'],
     'type': 'object'}

On instance:
    {'status': 200}
[ FAILED ]
```

I think it's because the conformance suite has a wrong JSON schema of the reply; there's no `"system_setting_key"` in the actual messages. You can try my suggested fix in this commit: https://github.com/device-automation-bus/dab-compliance-suite/commit/8a953a19ccf05986bbdd0cd085e8fef1fe9a3604.

**Comment by bgy36ww** [comment]
> On top of that validation of `/system/settings/set` responses is failing:
> 
> ```
> testing system/settings/set   {"memc": true} ... [ 
> Error: self.dab_client.last_error_code()
> Request invalid or malformed ]
> 
> testing system/settings/set   {"cec": true} ... Schema error: 'system_setting_key' is a required property
> 
> Failed validating 'required' in schema:
>     {'properties': {'error': {'type': ['string', 'null']},
>                     'status': {'type': 'integer'},
>                     'system_setting_key': {'type': 'string'},
>                     'value': {}},
>      'required': ['status', 'system_setting_key', 'value'],
>      'type': 'object'}
> 
> On instance:
>     {'status': 200}
> [ FAILED ]
> ```
> 
> I think it's because the conformance suite has a wrong JSON schema of the reply; there's no `"system_setting_key"` in the actual messages. You can try my suggested fix in this commit: [8a953a1](https://github.com/device-automation-bus/dab-compliance-suite/commit/8a953a19ccf05986bbdd0cd085e8fef1fe9a3604).

https://github.com/device-automation-bus/dab-specification-2.0?tab=readme-ov-file#operation-setting-system-settings
Oh this part I am confused. WDYM?

**Comment by xhaakon** [comment]
On 13. 02. 24 0:38, Wei Wang wrote:
> https://github.com/device-automation-bus/dab-specification-2.0?tab=readme-ov-file#operation-setting-system-settings 
> Oh this part I am confused. WDYM?

I agree this request format isn't defined very clearly:

interface SetSystemSettingsRequest extends DabRequest {
    [system_setting_key: keyof SystemSettings]: [value: any];
}

First of all I think instead of SystemSettings (which isn't defined anywhere in DAB spec) it should rather
reference GetCurrentSystemSettingsResponse inferface, like this:

interface SetSystemSettingsRequest extends DabRequest {
    [system_setting_key: keyof GetCurrentSystemSettingsResponse]: [value: any];
}

where [system_setting_key: keyof GetCurrentSystemSettingsResponse] says that SetSystemSettingsRequest
can contain properties of the same names as in GetCurrentSystemSettingsResponse. It's a more succinct
(and, arguably, confusing) way of writing:

interface SetSystemSettingsRequest extends DabRequest {
    language?: rfc_5646_language_tag;
    outputResolution?: OutputResolution;
    memc?: boolean;
    cec?: boolean;
    lowLatencyMode?: boolean;
    matchContentFrameRate?: MatchContentFrameRate;
    hdrOutputMode?: HdrOutputMode;
    pictureMode?: PictureMode;
    audioOutputMode?: AudioOutputMode;
    audioOutputSource?: AudioOutputSource;
    videoInputSource?: VideoInputSource;
    audioVolume?: int;
    mute?: boolean;
    textToSpeech?: boolean;
}

An example /system/settings/set request that changes outputResolution and audioVolume could then look like this:

{
    "outputResolution": {
       "width": 3840,
       "height": 2160,
       "frequency": 60
    },
    "audioVolume": 50
}

File: schema.py [hunk 5]
```diff
@@ -273,44 +273,75 @@
...
+        "properties": system_settings_schema
+    },
+    "else": {
+        "properties": {
+            "error": {"type": ["string", "null"]}
+        },
+        "required": ["error"]
+    },
+
+    "unevaluatedProperties": False
```
Comment (xhaakon) [inline schema.py:344]: Missing closing curly bracket
```suggestion
    "unevaluatedProperties": False
}
```

**Review by xhaakon — CHANGES_REQUESTED** [review]

**Comment by GaganBhat** [comment]
Agreed @xhaakon, the set systems request should only be able to set a system setting that the device is capable of getting in the first place.

However, when certifying we expect the DAB device to respond to all system setting options (even if it doesn't support it) by responding appropriately (501 not implemented) with any message of their choosing so that the client understands the reason for why their set failed.

**Comment by GaganBhat** [comment]
Additionally, I'm not sure if this PR consists of this:

But we need to verify that after doing a `set` operation, the `get` of that same property reflects the change as we expect.

We're seeing on some partner platforms that they return `200 successful` on a `set`, but once you `get` that same property it's not reflected correctly.

This will catch issues when the partner implementations for set system settings does not work as expected.

**Comment by bgy36ww** [comment]
> Additionally, I'm not sure if this PR consists of this:
> 
> But we need to verify that after doing a `set` operation, the `get` of that same property reflects the change as we expect.
> 
> We're seeing on some partner platforms that they return `200 successful` on a `set`, but once you `get` that same property it's not reflected correctly.
> 
> This will catch issues when the partner implementations for set system settings does not work as expected.

That would require a functional test. I think let's get the basic topic calls merge first.

**Comment by bgy36ww** [comment]
@xhaakon Need your approval.

File: schema.py [hunk 6]
```diff
@@ -273,55 +273,75 @@
...
+    "required": ["status"],
+
+    "if": {
+        "properties": {
+            "status": { "const": 200 }
+        }
+    },
+    "then": {
+        "properties": system_settings_schema,
+        "required": system_settings_schema.keys()
```
Comment (xhaakon) [inline schema.py:309]: This is my mistake here; `dict.keys()` returns a "view object", whereas jsonschema expects an array type as value of `"required"`. The suggestion below fixes this error:
```
Schema error: dict_keys(['language', 'outputResolution', 'memc', 'cec', 'lowLatencyMode', 'matchContentFrameRate', 'hdrOutputMode', 'pictureMode', 'audioOutputMode', 'audioOutputSource', 'videoInputSource', 'audioVolume', 'mute', 'textToSpeech']) is not of type 'array'
```

```suggestion
        "required": list(system_settings_schema.keys())
```

**Review by xhaakon — CHANGES_REQUESTED** [review]
I found one more issue (in my own code, sorry). Otherwise LGTM.

**Comment by bgy36ww** [comment]
Need LGTM from someone.

**Pull request merged**